addons/delivery_shipper/models/delivery_carrier.py

```diff
@@ -0,0 +1,245 @@
+                selected_rate_type.lower() == rate['rate_type'].lower()
+            )
+        ]
+        if not filtered_rates:
+            raise UserError(_('No rates can be found. Please check the package dimension or Shipper configuration'))
+
+        if filtered_rates[0]:
+            if filtered_rates[0]['rate_id'] != selected_id:
+                self.shipper_3pl_partner_id = filtered_rates[0]['rate_id']
+
+            return {
+                'success': True,
+                'price': filtered_rates[0].get('final_price'),
+                'error_message': False,
+                'warning_message': False,
+            }
+        else:
+            return {
+                'success': False,
+                'price': 0.0,
+                'error_message': _('Error: this delivery method is not available for this order.'),
+                'warning_message': False,
+            }
+
+    def get_shipper_rate(self, order):
+        """ Get the rates for the given order, according to the selected service code for this carrier.
+        This method is used when getting the rate for a specific shipping.method.
+        """
+        shipper = self._get_shipper()
+        origin_area_id = self._shipper_get_area_id(origin=self.shipper_origin_address)
+        destination_area_id = self._shipper_get_area_id(order=order)
+
+        rates = shipper._rate_shipment(
+            self._shipper_get_package_information(order=order),
+            origin_area_id,
+            destination_area_id,
+            order=order,
+        )
+
+        return rates
+
+    # API HELPERS #
+    def _shipper_get_area_id(self, order=False, origin=False):
+        shipper = self._get_shipper()
+        partner = order.partner_id if order and order.partner_id else origin
+
+        res = shipper._get_area_id(partner)
+        for area in res.get('data', []):
+            adm_level_5 = area.get('adm_level_5', {})
+
+            if (adm_level_5.get('postcode') == partner.zip):
+                return adm_level_5.get('id')
+        return False
+
+    def _shipper_get_package_information(self, order=False, picking=False):
+        original_weight_uom = self.env['product.template'].sudo()._get_weight_uom_id_from_ir_config_parameter()
+        target_weight_uom = self.env.ref('uom.product_uom_kgm')
+
+        total_weight, total_volume = 0, 0
+        if order:
+            for line in order.order_line.filtered(lambda line: not line.is_delivery and not line.display_type):
+                total_weight += line.product_uom_qty * line.product_id.weight
+                total_volume += line.product_uom_qty * line.product_id.volume
+        elif picking:
+            for move in picking.move_ids_without_package.filtered(lambda move: move.state not in ('done', 'cancel')):
+                total_weight += move.quantity * move.product_id.weight
+                total_volume += move.quantity * move.product_id.volume
```

Comment:
You should use the get package method and compute the information based on the packages, not the products.
Otherwise your rate will be wrong.